### PR TITLE
add gpu metrics

### LINKED ISSUE #1449: GPU metrics on kubernetes
The GPU task will be supported on kubernetes soon.
I wonder how to add GPU metrics on heapster, so the users can view the usage of GPU information via Grafana+InfluxDB.

### PR DESCRIPTION
**Pull request by andyxning**
This PR partially fix PR #1872  .

This PR adds support for scrape GPU related metrics when using normal cAdvisor endpoint instead of kubelet summary api. 

* `accelerator/memory_total`
* `accelerator/memory_used`
* `accelerator/duty_cycle`

are added as new accelerator metrics.

/cc @DirectXMan12 @loburm @piosz @mindprince #

### CODE REVIEW

**Comment by mindprince** [comment]
Why use the cAdvisor endpoint instead of the summary API?

Also, when we expose these metrics, we would probably expose them as
`accelerator/...`

**Comment by andyxning** [comment]
@mindprince 

> Why use the cAdvisor endpoint instead of the summary API?

For now at least i use the original cAdvisor api to collect metrics. I also think that there are some people use normal cAdvisor endpoint instead of summary endpoint. :)

>when we expose these metrics, we would probably expose them as
accelerator/...

Done. PTAL.

**Comment by piosz** [comment]
LGTM from me. I'll merge it on @mindprince's LGTM.

BTW is anyone going to add support for this in Summary API?

File: docs/storage-schema.md
Comment (mindprince) [inline docs/storage-schema.md]: Use the names that are used in the upstream API.
```
accelerator/memory_total
accelerator/memory_used
accelerator/duty_cycle
```

File: docs/storage-schema.md
Comment (mindprince) [inline docs/storage-schema.md]: Let's not use `nvidia` in the metric name and instead use labels if possible.

File: metrics/core/metrics.go
Comment (mindprince) [inline metrics/core/metrics.go]: Instead of Utility use DutyCycle. They are different things.

File: metrics/core/metrics.go
Comment (mindprince) [inline metrics/core/metrics.go]: MemoryTotal

File: metrics/processors/node_autoscaling_enricher.go
Comment (mindprince) [inline metrics/processors/node_autoscaling_enricher.go]: ResourceNvidiaGPU is a deprecated resource and is going away. The new resources are exposed through device plugins. I would not add this metric at all for now.

File: metrics/core/metrics.go
Comment (mindprince) [inline metrics/core/metrics.go]: Am I reading this right? This is just one label with key=LabelResourceID.Key and value=concatenation of Make, Model and ID? If so, I don't think we want this.

All these metrics should have three extra labels.
make, model and accelerator_id.

**Review by rohitagarwal003 — CHANGES_REQUESTED** [review]
You will need to add tests.

**Comment by mindprince** [comment]
@piosz I took a quick look and left some comments.

I also have a question if we even want to do this or just want to do through the summary API? Can both co-exist?

File: metrics/core/metrics.go
Comment (andyxning) [reply metrics/core/metrics.go]: Will do.

File: metrics/processors/node_autoscaling_enricher.go
Comment (andyxning) [reply metrics/processors/node_autoscaling_enricher.go]: > ResourceNvidiaGPU is a deprecated resource and is going away. The new resources are exposed through device plugins.

Does the resources has a predefined or patterned names? @mindprince

Comment (andyxning) [reply metrics/processors/node_autoscaling_enricher.go]: As for now, It seems that [`alpha.kubernetes.io/nvidia-gpu` is still available](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L3840). So, i think it is fine to add this for now. 

WDYT @piosz @mindprince

File: metrics/core/metrics.go
Comment (andyxning) [reply metrics/core/metrics.go]: Will do.

File: metrics/core/metrics.go
Comment (andyxning) [reply metrics/core/metrics.go]: Will do.

File: docs/storage-schema.md
Comment (andyxning) [reply docs/storage-schema.md]: Will do.

File: metrics/processors/node_autoscaling_enricher.go
Comment (andyxning) [reply metrics/processors/node_autoscaling_enricher.go]: I have add a test for `ResourceNvidiaGPU` resource name, so even it is not supported and removed in the future, everything should work fine.

File: docs/storage-schema.md
Comment (andyxning) [reply docs/storage-schema.md]: Will do.

**Comment by andyxning** [comment]
@mindprince @piosz @DirectXMan12 comments addressed. PTAL.

**Comment by andyxning** [comment]
@mindprince 
> I also have a question if we even want to do this or just want to do through the summary API? Can both co-exist?

normal api and summary api should both be supported, IMO. And, they can co-exist.

**Comment by andyxning** [comment]
Pring @mindprince

**Comment by DirectXMan12** [comment]
Hmm... I'd very much just like to say "always use the summary API", and just deprecate the "legacy" stats API usage entirely.  WDYT, @piosz @kubernetes/heapster-maintainers ?

**Comment by mindprince** [comment]
Agreed. And these were added to the summary API in #1891

**Comment by andyxning** [comment]
@DirectXMan12 @mindprince 

Until kubernetes disable cAdvisor integration, we should support `legacy` stats API usage. Just like what we have discussed in #1881 , this deprecation of `legacy` stats API is depends on the kubernetes. And, for this time i agree that we should make an deprecation statement along side kubernetes release on which cAdvisor integration is disabled.

**Comment by andyxning** [comment]
@DirectXMan12 @loburm @mindprince Please take a look at this PR. Thanks.

File: metrics/core/labels.go
Comment (mindprince) [inline metrics/core/labels.go]: Get rid of this diff. It's changing the labels from what I added in the other PR.

File: metrics/core/labels.go
Comment (mindprince) [inline metrics/core/labels.go]: Adding these labels here adds these labels to a lot of metrics where they don't make sense at all. You should use `acceleratorLabels` instead.

Comment (mindprince) [reply metrics/core/labels.go]: All the diff from this file can go away.

File: metrics/core/metrics.go
Comment (mindprince) [inline metrics/core/metrics.go]: Keep this.

File: metrics/processors/node_autoscaling_enricher.go
Comment (mindprince) [inline metrics/processors/node_autoscaling_enricher.go]: As I said before, this is deprecated and the resource management WG doesn't want to add more code that depends on this. Please get rid of this metric completely.

**Review by rohitagarwal003 — CHANGES_REQUESTED** [review]
I am not sure what the final decision would be in terms of whether we want to add this at all or not.

But I took another pass and left some comments.

File: metrics/core/labels.go
Comment (andyxning) [reply metrics/core/labels.go]: I think it is a good thing to make all accelerator related labels `accelerator` prefixed. This may be useful for classification.

File: metrics/core/labels.go
Comment (andyxning) [reply metrics/core/labels.go]: Will do.

File: metrics/core/metrics.go
Comment (andyxning) [reply metrics/core/metrics.go]: Will do.

File: metrics/processors/node_autoscaling_enricher.go
Comment (andyxning) [reply metrics/processors/node_autoscaling_enricher.go]: Will do.

**Comment by andyxning** [comment]
@mindprince All comments are addressed. PTAL.

File: metrics/core/labels.go
Comment (mindprince) [reply metrics/core/labels.go]: These labels only show up for accelerator metrics (because we use `acceleratorLabels` only in accelerator metrics). Because of this, I think it is better to not repeat accelerator in the label name. `accelerator_id` is a special case because `id` is usually used for other things.

Also, this will keep things consistent with Stackdriver and prometheus.

File: metrics/core/metrics.go
Comment (mindprince) [reply metrics/core/metrics.go]: Minor nit: if you move this field from last to here, the diff would be shorter.

Comment (andyxning) [reply metrics/core/metrics.go]: Will do.

**Comment by andyxning** [comment]
@mindprince Comments addressed.

**Comment by andyxning** [comment]
Ping @piosz @DirectXMan12

**Comment by DirectXMan12** [comment]
> Until kubernetes disable cAdvisor integration, we should support legacy stats API usage or otherwise this is also a backward breaking change

Whether or not Kubernetes has cAdvisor integration is independent of whether we support the legacy stats API.  We can certainly choose to deprecate it and/or not add new features.  Ceasing to add new features is *not* a breaking change.

My vote would be to simply not add new features to our legacy API support, and mark it as deprecated.

**Comment by andyxning** [comment]
@DirectXMan12 What is the deprecation schedule for legacy stats api?

Ping @piosz @kubernetes/heapster-maintainers I want to hear more voice on this.

**Comment by andyxning** [comment]
@piosz Could you please take a look at this?

**Pull request merged**
